### PR TITLE
fix: calling of X11 functions when running under Wayland

### PR DESCRIPTION
**Pull request by p2004a**
#### Description of Change

**TLDR**: I primarily try to address the #33325 and #32436 issues with this PR.

In #30814 there was https://github.com/electron/electron/pull/30814/commits/fe4008832737f862cae3dad17e11b205ef177972 committed that removed all `if (!features::IsUsingOzonePlatform()) {` blocks. Those blocks were guarding calls to X11 functions under wayland and were added as part of https://github.com/electron/electron/pull/26022. Calling X11 functions under wayland can cause the application to crash (as visible in #33325 and #32436).

This PR:
- Replaces `USE_X11` with `USE_OZONE` blocks as ozone is the only build of chrome nowadays (Ref: https://github.com/electron/electron/issues/32436#issuecomment-1032338114) and only call X11 code only when running under X11. I believe the functionality hidden on non-x11 platform is not critical and this is overall net positive to current state.
- Replaces `WindowStateWatcher` with ozone native implementation in `ElectronDesktopWindowTreeHostLinux`: it is always instantiated under ozone, but the wayland decorations remain guarded by the `WaylandWindowDecorations` feature flag (As was added in #29618).
  
  This makes the `maximize`, `enter-full-screen`, etc. events of [`BrowserWindow`](https://www.electronjs.org/docs/latest/api/browser-window) work under wayland.
- Eliminates the X11 calls in file and message dialogs. Frankly I'm not sure if call to `gtk_window_present` is needed at all: under my system it's noop, but maybe there are DEs where it matters, so keeping it.

#### Checklist
- [X] PR description included and stakeholders cc'd
- [X] `npm test` passes: [kind of](https://github.com/electron/electron/pull/33355#issuecomment-1075660218) ~I can't make the tests even start running [1]~
- [X] [PR release notes](https://github.com/electron/clerk/blob/master/README.md) describe the change in a way relevant to app developers, and are [capitalized, punctuated, and past tense](https://github.com/electron/clerk/blob/master/README.md#examples).

#### Release Notes

notes: Fixed multiple issues when running under Wayland caused by calling X11 functions.

~[1]: clean `e init --bootstrap testing` followed by `e test` I failing for me with `Running "ninja -j 200 node:headers" in /home/p2004a/Workspace/electron/e/src/out/Testing
ninja: error: unknown target 'node:headers', did you mean 'node_headers_1'?`, I can't make it work :(, I was testing functions with my own draft app. I hope CI run as part of this PR will handle this omission.~

### CODE REVIEW

**Comment by codebytere** [comment]
This also partially addresses https://github.com/electron/electron/issues/31382

**Review by dsanders11 — COMMENTED** [review]
Added some comments. I haven't tested this, just adding some suggested changes.

> [1]: clean e init --bootstrap testing followed by e test I failing for me with Running "ninja -j 200 node:headers" in /home/p2004a/Workspace/electron/e/src/out/Testing ninja: error: unknown target 'node:headers', did you mean 'node_headers_1'?, I can't make it work :(, I was testing functions with my own draft app. I hope CI run as part of this PR will handle this omission.

Thanks for mentioning that! Fixing it over in https://github.com/electron/build-tools/pull/351.

File: shell/browser/ui/file_dialog_gtk.cc [hunk 1]
```diff
@@ -247,14 +243,9 @@ class FileChooserDialog {
...
-#if defined(USE_X11)
-      // We need to call gtk_window_present after making the widgets visible
-      // to make sure window gets correctly raised and gets focus.
-      x11::Time time = ui::X11EventSource::GetInstance()->GetTimestamp();
-      gtk_window_present_with_time(GTK_WINDOW(dialog_),
-                                   static_cast<uint32_t>(time));
-#endif
+      // We need to call gtk_window_present after making the widgets visible to
+      // make sure window gets correctly raised and gets focus.
+      gtk_window_present(GTK_WINDOW(dialog_));
```
[the remark this answers is not in the document]
Comment (p2004a) [reply shell/browser/ui/file_dialog_gtk.cc]: Done

File: shell/browser/ui/message_box_gtk.cc [hunk 2]
```diff
@@ -161,14 +157,9 @@ class GtkMessageBox : public NativeWindowObserver {
...
     gtk_widget_show(dialog_);
-
-#if defined(USE_X11)
     // We need to call gtk_window_present after making the widgets visible to
     // make sure window gets correctly raised and gets focus.
-    x11::Time time = ui::X11EventSource::GetInstance()->GetTimestamp();
-    gtk_window_present_with_time(GTK_WINDOW(dialog_),
-                                 static_cast<uint32_t>(time));
-#endif
+    gtk_window_present(GTK_WINDOW(dialog_));
```
[the remark this answers is not in the document]
Comment (p2004a) [reply shell/browser/ui/message_box_gtk.cc]: Done

File: shell/browser/native_window_views.cc [hunk 3]
```diff
@@ -1169,7 +1179,10 @@ void NativeWindowViews::SetMenu(ElectronMenuModel* menu_model) {
   }
 
   // Use global application menu bar when possible.
-  if (ShouldUseGlobalMenuBar()) {
+  if (ui::OzonePlatform::GetInstance()
+          ->GetPlatformProperties()
+          .supports_global_application_menus &&
```
[the remark this answers is not in the document]
Comment (p2004a) [reply shell/browser/native_window_views.cc]: I hope done.

File: shell/browser/ui/electron_desktop_window_tree_host_linux.h [hunk 4]
```diff
@@ -64,6 +67,8 @@ class ElectronDesktopWindowTreeHostLinux
                           &views::LinuxUI::AddDeviceScaleFactorObserver,
                           &views::LinuxUI::RemoveDeviceScaleFactorObserver>
       scale_observation_{this};
+  bool wayland_window_decorations_;
```
[the remark this answers is not in the document]
Comment (p2004a) [reply shell/browser/ui/electron_desktop_window_tree_host_linux.h]: Good point, not sure why I written it this way..

**Comment by p2004a** [comment]
I've rebased the branch because there was a merge conflict.

Thank you @dsanders11 for taking a look and fixing build tools!

I've run the tests and I've got exactly the same result on the `origin:main` and on my `p2004a:wayland-fixes` branches so I don't think I've broken anything ;). It doesn't mean that tests pass though, 10 tests don't, and those are the same tests on both main and my branch: https://gist.github.com/p2004a/480d7248de118c86c204c785bd05a9fd. I don't plan on debugging those here.

File: shell/browser/native_window_views.cc [hunk 5]
```diff
@@ -157,6 +162,10 @@ NativeWindowViews::NativeWindowViews(const gin_helper::Dictionary& options,
                                      NativeWindow* parent)
     : NativeWindow(options, parent),
       root_view_(std::make_unique<RootView>(this)),
+      // It's hacky to use GetPlatformNameForTest to figure out if we run under
+      // X11 but that's the best worst way at the moment.
+      is_x11_(ui::OzonePlatform::GetInstance()->GetPlatformNameForTest() ==
+              "x11"),
```
Comment (zcbenz) [inline shell/browser/native_window_views.cc:168]: Should be wrapped in `#if defined(USE_OZONE)`.

File: shell/browser/ui/file_dialog_gtk.cc [hunk 6]
```diff
@@ -18,12 +18,9 @@
 #include "shell/browser/unresponsive_suppressor.h"
 #include "shell/common/gin_converters/file_path_converter.h"
 #include "ui/base/glib/glib_signal.h"
+#include "ui/gtk/gtk_ui.h"
```
Comment (zcbenz) [inline shell/browser/ui/file_dialog_gtk.cc:21]: This is generating a GN warning:

```
ERROR at //electron/shell/browser/ui/message_box_gtk.cc:21:11: Including a private header.
#include "ui/gtk/gtk_ui.h"
          ^--------------
This file is private to the target //ui/gtk:gtk
```

I'm not sure what is the best way to fix it. /cc @nornagon

File: shell/browser/native_window_views.cc [hunk 7]
```diff
@@ -157,6 +162,10 @@ NativeWindowViews::NativeWindowViews(const gin_helper::Dictionary& options,
                                      NativeWindow* parent)
     : NativeWindow(options, parent),
       root_view_(std::make_unique<RootView>(this)),
+      // It's hacky to use GetPlatformNameForTest to figure out if we run under
+      // X11 but that's the best worst way at the moment.
+      is_x11_(ui::OzonePlatform::GetInstance()->GetPlatformNameForTest() ==
```
Comment (deepak1556) [inline shell/browser/native_window_views.cc:167]: We generally avoid testing apis as they are prone to removal, I think this can be replaced with `#if BUILDFLAG(OZONE_PLATFORM_X11)`

File: shell/browser/native_window_views.cc [hunk 8]
```diff
@@ -130,6 +125,16 @@ gfx::Rect DIPToScreenRect(HWND hwnd, const gfx::Rect& pixel_bounds) {
 
 #endif
 
+#if defined(USE_OZONE)
+
+bool CreateGlobalMenuBar() {
```
Comment (zcbenz) [inline shell/browser/native_window_views.cc:130]: A function named `CreateGlobalMenuBar` usually means it is used to create the menu bar, I think we should move the testing of `supports_global_application_menus` to `ShouldUseGlobalMenuBar`.

File: shell/browser/ui/file_dialog_gtk.cc [hunk 9]
```diff
@@ -18,12 +18,9 @@
...
 #include "ui/gtk/gtk_util.h"
 
-#if defined(USE_X11)
-#include "ui/events/platform/x11/x11_event_source.h"
-#endif
-
 #if defined(USE_OZONE) || defined(USE_X11)
 #include "ui/base/ui_base_features.h"
```
Comment (deepak1556) [inline shell/browser/ui/file_dialog_gtk.cc:25]: Unused header, can be removed.

**Review by deepak1556 — COMMENTED** [review]
Thanks for working on this, really appreciate it! 

Reviews are based on first pass on the PR,

1) Can you also remove the `USE_X11` define from https://github.com/electron/electron/blob/8ad1470d08435de405b6738ef546d5ca95559fd0/BUILD.gn#L562 

2) Can you apply the changes from https://chromium-review.googlesource.com/c/chromium/src/+/2289772/ and https://chromium-review.googlesource.com/c/chromium/src/+/2317185 to `GlobalMenuBarX11` which better aligns with upstream. Should end up in a state similar to https://source.chromium.org/chromium/chromium/src/+/main:chrome/browser/ui/BUILD.gn;l=3873-3882

File: shell/browser/ui/file_dialog_gtk.cc (hunk 6 above)
Comment (deepak1556) [reply shell/browser/ui/file_dialog_gtk.cc:21]: Wonder if we can refactor `file_dialog_gtk` to reuse https://source.chromium.org/chromium/chromium/src/+/main:ui/gtk/select_file_dialog_linux_gtk.h via `//ui/shell_dialogs` . We already started using this component in other places https://github.com/electron/electron/pull/30663

File: shell/browser/native_window_views.cc (hunk 8 above)
Comment (dsanders11) [reply shell/browser/native_window_views.cc:130]: @zcbenz, I actually [suggested he add this](https://github.com/electron/electron/pull/33355#discussion_r831707161), since it's from upstream. Agreed on the name being confusing, but that's what it's called in upstream. I was hoping to keep any new code here aligned with upstream so it'll be easier to spot divergences.

Comment (zcbenz) [reply shell/browser/native_window_views.cc:130]: Thanks for clarification, it makes sense then 👍 .

File: shell/browser/ui/file_dialog_gtk.cc (hunk 6 above)
Comment (dsanders11) [reply shell/browser/ui/file_dialog_gtk.cc:21]: @zcbenz, looks like there's [already a patch](https://github.com/electron/electron/blob/3da598015bccb8f9ab6833d1163c2fe2b0f2b058/patches/chromium/ui_gtk_public_header.patch) exposing "ui/gtk/gtk_util.h", so we could just tack on this header there as well?

@deepak1556, I think we'd still need to expose the header for the usage in `message_box_gtk`, don't see an upstream class we could reuse for that.

Comment (deepak1556) [reply shell/browser/ui/file_dialog_gtk.cc:21]: Yeah but it is only a single utility `gtk::SetGtkTransientForAura` we are relying on from that header, maybe we can inline it. I am good with the header patching for unblocking this PR, the refactor can be explored as follow-up.

File: shell/browser/native_window_views.cc (hunk 7 above)
Comment (p2004a) [reply shell/browser/native_window_views.cc:167]: I don't think we can guard on build flag. Running under X11 is a runtime property, the same binary can run under wayland, X11, or other platforms supported by ozone layer.

Alternative approach here would be to use [`skia_can_fall_back_to_x11`](https://source.chromium.org/chromium/chromium/src/+/main:ui/ozone/public/ozone_platform.h;bpv=1;bpt=1;l=133?q=GetPlatformNameForTest&ss=chromium&gsn=skia_can_fall_back_to_x11&gs=kythe%3A%2F%2Fchromium.googlesource.com%2Fchromium%2Fsrc%3Flang%3Dc%252B%252B%3Fpath%3Dsrc%2Fui%2Fozone%2Fpublic%2Fozone_platform.h%23cr2mt2XFdJF2t4pnh2ZuoaoFuKd13NmSrxYfYoW67zs) or add a patch for upstream that introduces new dedicated platform property like `electron_can_call_x11`

File: shell/browser/ui/file_dialog_gtk.cc (hunk 9 above)
Comment (p2004a) [reply shell/browser/ui/file_dialog_gtk.cc]: Done

File: shell/browser/native_window_views.cc (hunk 5 above)
Comment (p2004a) [reply shell/browser/native_window_views.cc:169]: Oops, fixed!

File: shell/browser/ui/file_dialog_gtk.cc (hunk 6 above)
Comment (zcbenz) [reply shell/browser/ui/file_dialog_gtk.cc:21]: > looks like there's [already a patch](https://github.com/electron/electron/blob/3da598015bccb8f9ab6833d1163c2fe2b0f2b058/patches/chromium/ui_gtk_public_header.patch) exposing "ui/gtk/gtk_util.h", so we could just tack on this header there as well?

Yeah adding the header to the patch would work.

File: shell/browser/native_window_views.cc (hunk 7 above)
Comment (deepak1556) [reply shell/browser/native_window_views.cc:168]: Ah okay, makes sense. In that case why not use the non-test variant [ui::GetOzonePlatformName](https://source.chromium.org/chromium/chromium/src/+/main:ui/ozone/platform_selection.h;l=17) ? Also I would still consider the build guards to use `BUILDFLAG(OZONE_PLATFORM_X11)` for X11 paths so that we allow configuring `ozone_platform_*` gn args.

Comment (p2004a) [reply shell/browser/native_window_views.cc:168]: `ui/ozone/platform_selection.h` is a private header, that would need another patch to build files to make it public.

Using `BUILDFLAG(OZONE_PLATFORM_X11)` for X11 paths sounds good, I will send commit latter to do that.

Comment (deepak1556) [reply shell/browser/native_window_views.cc:168]: [//ui/ozone](https://source.chromium.org/chromium/chromium/src/+/main:ui/ozone/BUILD.gn;l=244-251) component is built with public visibility, so it should be possible to include headers from there without any patch.

Comment (p2004a) [reply shell/browser/native_window_views.cc:168]: Sure, the component is public, but [the header file is private](https://source.chromium.org/chromium/chromium/src/+/main:ui/ozone/BUILD.gn;drc=62fc260067cf0c17d5416448675bfd73dbf6db5a;l=183), [those are the public headers](https://source.chromium.org/chromium/chromium/src/+/main:ui/ozone/BUILD.gn;drc=62fc260067cf0c17d5416448675bfd73dbf6db5a;l=192-194).

To make the point: when I add this include, `e d gn check out/Testing //electron:electron_lib` returns:
```
ERROR at //electron/shell/browser/native_window_views.cc:67:11: Including a private header.
#include "ui/ozone/platform_selection.h"
          ^----------------------------
This file is private to the target //ui/ozone:platform
```

File: shell/browser/ui/file_dialog_gtk.cc (hunk 6 above)
Comment (p2004a) [reply shell/browser/ui/file_dialog_gtk.cc:21]: I've added the `gtk_ui.h` header to the patch which I believe resolves this thread.

File: shell/browser/native_window_views.cc (hunk 7 above)
Comment (p2004a) [reply shell/browser/native_window_views.cc]: If we don't want to use test functions, and I will have to patch something, I've wend down the path of IMHO the cleanest solution and introduced `electron_can_call_x11` property.

Comment (deepak1556) [reply shell/browser/native_window_views.cc]: We actually include private headers when needed in the `//electron_lib` target, you can grep for `// nogncheck` in the code. If a target is visible then it should not be an issue to use the private headers, as long as it is ensured that the target is built. Now that I think about it, we could have also added the `//  nogncheck` for `gtk_util.h` header instead of the patch, will look into that in a follow-up.

> I've wend down the path of IMHO the cleanest solution and introduced electron_can_call_x11 property.

I like this 👍

File: shell/browser/ui/electron_desktop_window_tree_host_linux.cc [hunk 10]
```diff
@@ -46,13 +48,53 @@ void ElectronDesktopWindowTreeHostLinux::OnBoundsChanged(
     const BoundsChange& change) {
   views::DesktopWindowTreeHostLinux::OnBoundsChanged(change);
   UpdateFrameHints();
+  // The OnWindowStateChanged should receive all updates but currently it
+  // doesn't receive changes to fullscreen status, so we catch them here.
+  UpdateWindowState(platform_window()->GetPlatformWindowState());
```
Comment (deepak1556) [inline shell/browser/ui/electron_desktop_window_tree_host_linux.cc:53]: I tested the branch and this only seems to affect X11 platform, stems from [synchronously handling the fullscreen states](https://source.chromium.org/chromium/chromium/src/+/main:ui/ozone/platform/x11/x11_window.cc;l=592-605) and ends up not firing the [state change](https://source.chromium.org/chromium/chromium/src/+/main:ui/ozone/platform/x11/x11_window.cc;l=1353-1357).

Atleast in my testing wayland platform does not suffer from this, can we guard this behind `electron_can_call_x11` and also improve the comments around it.

**Review by deepak1556 — COMMENTED** [review]
LGTM with minor change.

File: shell/browser/ui/electron_desktop_window_tree_host_linux.cc (hunk 10 above)
Comment (p2004a) [reply shell/browser/ui/electron_desktop_window_tree_host_linux.cc]: Thanks for finding the code responsible! I've guarded code behind check and expanded comment.

**Review by deepak1556 — APPROVED** [review]
@p2004a can you rebase to resolve the conflicts, thanks!

**Comment by p2004a** [comment]
@deepak1556 Done!

**Review by deepak1556 — APPROVED** [review]

**Review by jkleinsc — APPROVED** [review]

**Comment by prusnak** [comment]
ISTM this still has to be backported to 16.x.x

**Comment by deepak1556** [comment]
The PR has not been backported to `16-x-y` yet, I didn't spend time on it because `16-x-y` also requires the following https://github.com/electron/electron/pull/32603 for the runtime to work on wayland. If anyone is interested to work on both these backports, feel free to!
